### PR TITLE
Add windows picker

### PR DESCRIPTION
**Pull request by kaiguo**
# Summary

Added Windows Picker.

Close https://github.com/microsoft/react-native-windows/issues/3036.
Close https://github.com/microsoft/react-native-windows/issues/2907.

### What's required for testing (prerequisites)?
0. [System requirments](https://microsoft.github.io/react-native-windows/docs/rnw-dependencies).
1. Run `yarn start:windows`.
2. Open `example/windows/PickerExample.sln` with VisualStudio and run.

picker-windows

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| Windows     |    ✅     |

## Checklist

- [X] I have tested this on a device and a simulator
- [X] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [x] I updated the typed files (TS and Flow)
- [x] I added a sample use of the API in the example project (`example/App.js`)

### CODE REVIEW

File: windows/ReactNativePicker/ReactPackageProvider.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿#include "pch.h"
+#include "ReactPackageProvider.h"
+#if __has_include("ReactPackageProvider.g.cpp")
+#include "ReactPackageProvider.g.cpp"
```
Comment (asklar) [inline windows/ReactNativePicker/ReactPackageProvider.cpp]: curious about when this file might/might not exist? what generates this?

File: windows/ReactNativePicker/ReactPickerView.cpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    using namespace Windows::UI::Xaml::Input;
+    using namespace Windows::UI::Xaml::Media;
+} // namespace winrt
+
+namespace winrt::ReactNativePicker::implementation {
+
+    ReactPickerView::ReactPickerView(winrt::IReactContext const& reactContext) : m_reactContext(reactContext) {
+        m_isEditableComboboxSupported = winrt::ApiInformation::IsPropertyPresent(
+            L"Windows.UI.Xaml.Controls.ComboBox", L"IsEditableProperty");
+        this->AllowFocusOnInteraction(true);
```
Comment (asklar) [inline windows/ReactNativePicker/ReactPickerView.cpp]: do we have other usages of WUX types? we should collect these somewhere to make it easier to migrate to WinUI3
e.g.
#define XAML_NAMESPACE  L"Windows.UI.Xaml"
#define MAKE_XAML_TYPE(x) XAML_NAMESPACE ## L"." ## L#x

then you can pass MAKE_XAML_TYPE(Controls.ComboBox)

when WinUI3 comes along we just change XAML_NAMESPACE to L"Microsoft.UI.Xaml"

File: windows/ReactNativePicker/ReactPickerView.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            if (auto self = ref.get()) {
+                self->OnSelectionChanged(sender, args);
+            }
+        });
+
+        m_dropDownClosedRevoker = this->DropDownClosed(winrt::auto_revoke,
+            [](auto const& sender, auto const& /*args*/) {
+                // When the drop down closes, attempt to move focus to its anchor
+                // textbox to prevent cases where focus can land on an outer flyout
+                // content and therefore trigger a unexpected flyout dismissal
```
Comment (asklar) [inline windows/ReactNativePicker/ReactPickerView.cpp]: consider re-wording? hard to parse this right now :)

File: windows/ReactNativePicker/ReactPickerView.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        m_updating = true;
+        bool updateSelectedIndex = false;
+        const JSValueObject& propertyMap = JSValueObject::ReadFrom(reader);
+
+        for (auto const& pair : propertyMap) {
+            auto const& propertyName = pair.first;
+            auto const& propertyValue = pair.second;
+
+            if (propertyName == "editable") {
+                if (m_isEditableComboboxSupported) {
```
Comment (asklar) [inline windows/ReactNativePicker/ReactPickerView.cpp]: meta-point / don't need to hold the PR for this

I see we do similar things in every module/control we implement:

we basically have a big if/else/else/... of all the RN property names we support, and we manually map those things to XAML properties one by one.
We have the metadata of XAML properties (types, indices, etc.) and we can build a metadata table to make all of this data driven instead of writing the same code over and over

File: windows/ReactNativePicker/ReactPickerView.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+    void ReactPickerView::SetSelectedIndex(int index) {
+        if (index < static_cast<int32_t>(this->Items().Size())) {
+            this->SelectedIndex(index);
+            // Keep comboBox.Foreground and selectedItem.Foreground in sync.
+            if (auto selectedItem = this->SelectedItem().try_as<winrt::ComboBoxItem>()) {
+                this->Foreground(selectedItem.Foreground());
+            }
+        }
+    }
```
Comment (asklar) [inline windows/ReactNativePicker/ReactPickerView.cpp]: else debug spew?

File: windows/ReactNativePicker/ReactPickerView.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    public:
+        ReactPickerView(Microsoft::ReactNative::IReactContext const& reactContext);
+        void UpdateProperties(Microsoft::ReactNative::IJSValueReader const& reader);
+
+    private:
+        Microsoft::ReactNative::IReactContext m_reactContext{ nullptr };
+        bool m_updating{ false };
+        // FUTURE: remove when we can require RS5+
+        bool m_isEditableComboboxSupported{ false };
+        int32_t m_selectedIndex{ 0 };
```
Comment (asklar) [inline windows/ReactNativePicker/ReactPickerView.h]: this should be a static, no need to have this on every instance

File: example/windows/PickerExample/Package.appxmanifest [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+<Package
+  xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
+  xmlns:mp="http://schemas.microsoft.com/appx/2014/phone/manifest"
+  xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
+  IgnorableNamespaces="uap mp">
+
+ <Identity
+    Name="e99f71c4-2760-471e-b3e9-8ebcf72fa94a"
+    Publisher="CN=kaigu"
+    Version="1.0.0.0" />
```
Comment (asklar) [inline example/windows/PickerExample/Package.appxmanifest]: update this an PublisherDisplayName?

File: windows/ReactNativePicker/pch.h [hunk 8]
```diff
@@ -1,6 +1,9 @@
 ﻿#pragma once
 
 #define NOMINMAX
+#define XAML_NAMESPACE L"Windows.UI.Xaml"
```
Comment (asklar) [inline windows/ReactNativePicker/pch.h]: if this could go into a header outside the specific picker control, that'd be great (not sure if we already have other controls in this repo?)

Comment (kaiguo) [reply windows/ReactNativePicker/pch.h]: This repo is picker only. Maybe we should add it somewhere in react-native-windows, perhaps Microsoft.ReactNative.Cxx.

File: windows/ReactNativePicker/ReactPackageProvider.cpp (hunk 1 above)
Comment (kaiguo) [reply windows/ReactNativePicker/ReactPackageProvider.cpp]: This was copied from the old PickerViewManager. I guess the file should always be there in current setup, removed the if statement.

File: windows/ReactNativePicker/ReactPickerView.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "ReactPickerView.g.h"
+#include "winrt/Microsoft.ReactNative.h"
+#include "NativeModules.h"
+
+namespace winrt::ReactNativePicker::implementation {
```
Comment (kmelmon) [inline windows/ReactNativePicker/ReactPickerView.h]: >ReactNativePicker [](start = 17, length = 17)

Question:  Does this implementation do the equivalent of supplying a custom yoga measure function like it currently does in react-native-windows?  Picker is special in that it can size itself to the selected text, see PickerViewManager::GetYogaCustomMeasureFunc()

Comment (kaiguo) [reply windows/ReactNativePicker/ReactPickerView.h]: @jonthysell (or Keith) what interface or callback should I implement to plug in our own measure function? The one (GetYogaCustomMeasureFunc) we used in react::uwp is an override of the function in [ViewManagerBase](https://github.com/microsoft/react-native-windows/blob/master/vnext/ReactUWP/Views/ViewManagerBase.cpp) which doesn't seem to be accessible here.

I looked at [IViewManager.idl](https://github.com/microsoft/react-native-windows/blob/master/vnext/Microsoft.ReactNative/IViewManager.idl), but didn't find anything related...

Comment (kmelmon) [reply windows/ReactNativePicker/ReactPickerView.h]: I don't believe we have any mechanism in place yet to handle this.  It doesn't seem too difficult though in principle, as the code in UWP just needs a XAML FrameworkElement to measure.  

I'm not sure how Picker behaves without this.  Does it require the app to set a width/height?  Ideally it works well enough for us to file a follow-up issue and addressing after this is checked in.

Comment (kaiguo) [reply windows/ReactNativePicker/ReactPickerView.h]: Ah I see, I was wondering why it didn't render without width/height, looks like that's the reason. I've added height in the default style so app author doesn't have to set width/height.

https://github.com/react-native-community/react-native-picker/blob/5260f4887137c3417d1d22f7350d8a3fcde7ab2a/js/PickerWindows.windows.js#L137-L140

I'll file a follow-up issue about the measure override.

Comment (kmelmon) [reply windows/ReactNativePicker/ReactPickerView.h]: thanks!

**Comment by NickGerleman** [comment]
Did this change mess up the line endings on existing files? Looks like some are wholesale replaced.

**Comment by kaiguo** [comment]
> Did this change mess up the line endings on existing files? Looks like some are wholesale replaced.

Yeah probably, let me see if I can fix it.

**Comment by kaiguo** [comment]
> Did this change mess up the line endings on existing files? Looks like some are wholesale replaced.

Should be fixed now.

**Comment by NickGerleman** [comment]
@kaiguo Sorry for the late flood of comments. I forgot to submit a CR a couple weeks ago 😬

File: js/Picker.js [hunk 10]
```diff
@@ -146,6 +147,12 @@ class Picker extends React.Component<PickerProps> {
          * was added when renaming suppression sites. */
         <PickerAndroid {...this.props}>{this.props.children}</PickerAndroid>
       );
+    } else if (Platform.OS === 'windows') {
+      return (
+        /* $FlowFixMe(>=0.81.0 site=react_native_android_fb) This suppression
```
[the remark this answers is not in the document]
Comment (kaiguo) [reply js/Picker.js]: Ah, it was copied from the block above, removed.

File: js/PickerWindows.windows.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+import type {SyntheticEvent} from 'CoreEventTypes';
+import type {TextStyleProp} from 'StyleSheet';
+
+const RCTPicker = requireNativeComponent('RCTPicker');
+
+type PickerWindowsChangeEvent = SyntheticEvent<
+  $ReadOnly<{|
+    nativeEvent: {
+      // tslint:disable-next-line:no-any
```
[the remark this answers is not in the document]
Comment (kaiguo) [reply js/PickerWindows.windows.js]: Removed.

File: js/PickerAndroid.windows.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+/**
```
[the remark this answers is not in the document]
Comment (kaiguo) [reply js/PickerAndroid.windows.js:1]: Updated PickerWindows, will update Android and iOS in a separate PR.

File: js/PickerWindows.windows.js [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+  NativeModules,
+  StyleSheet,
+} from 'react-native';
+
+const REF_PICKER = 'picker';
+
+import type {SyntheticEvent} from 'CoreEventTypes';
+import type {TextStyleProp} from 'StyleSheet';
+
+const RCTPicker = requireNativeComponent('RCTPicker');
```
Comment (NickGerleman) [inline js/PickerWindows.windows.js:23]: @kaiguo this will pull in the built in view manager as well, and will cause crashes in 0.62 if we include this due to duplicate registration of the same component.

File: js/PickerWindows.windows.js [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+      itemIndex: number,
+      text: string,
+    },
+  |}>,
+>;
+
+type PickerWindowsProps = $ReadOnly<{|
+  children?: React.Node,
+  style?: ?TextStyleProp,
+  // tslint:disable-next-line:no-any
```
Comment (NickGerleman) [inline js/PickerWindows.windows.js:38]: No tslint since this is Flow

File: js/PickerIOS.windows.js [hunk 15]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (NickGerleman) [inline js/PickerIOS.windows.js:2]: Remove Microsoft from the copyright headers?

File: js/PickerWindows.windows.js (hunk 13 above)
Comment (kaiguo) [reply js/PickerWindows.windows.js:23]: Yeah we should remove the built in version in 62. Or are you suggesting using a different module name here?

Comment (NickGerleman) [reply js/PickerWindows.windows.js:23]: It's not removed from upstream yet, only deprecated, so we shouldn't remove ours. We will want to rename this one. I would follow whatever the other platforms are doing for name here.

We also still need to support this on 0.61, and wouldn't want the two to point to the same view manager.

Comment (kaiguo) [reply js/PickerWindows.windows.js:23]: Makes sense, I'll rename it.

**Comment by chrisglein** [comment]
Was there something still blocking this PR?
@jainkuniya did you still have changes you were waiting for?
@kaiguo there are items still unchecked in the PR description - do those still need to be addressed?

**Comment by kaiguo** [comment]
> @kaiguo Kai Guo FTE there are items still unchecked in the PR description - do those still need to be addressed?

Just updated the list to reflect current state. The only one left is changelog file, which hasn't been updated for awhile so I guess I'll leave it for now.

**Comment by Naturalclar** [comment]
@kaiguo there's some minor conflicts due to #71, could you resolve the conflicts?

**Comment by kaiguo** [comment]
> @kaiguo Kai Guo FTE there's some minor conflicts due to #71, could you resolve the conflicts? Thanks!

@Naturalclar should be fixed now : )

**Comment by Naturalclar** [comment]
@jainkuniya Should be good to go.
Could you merge and release a new version when you get the chance?
